### PR TITLE
Add recursive CCS

### PR DESCRIPTION
**Pull request by FabienRoger**

### CODE REVIEW

File: elk/training/ccs.py [hunk 1]
```diff
@@ -102,7 +109,17 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
 
     def validate_data(self, data):
-        assert len(data) == 2 and data[0].shape == data[1].shape
+        assert (
+            len(data) == 2
+            and data[0].shape == data[1].shape
+            and data[0].dtype == data[1].dtype == self.dtype
+        ), "Data must be a tuple of two tensors of the same shape and dtype"
+
+    def correct_dtypes(
```
Comment (norabelrose) [inline elk/training/ccs.py:118]: Why is this necessary at all? I don't see why we need to cast to a single dtype

File: elk/extensions/recursive_ccs/train.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+from tqdm.auto import tqdm
+
+from elk.files import elk_cache_dir
+from elk.training.ccs import CCS
+from elk.training.preprocessing import load_hidden_states, normalize
+from elk.extensions.recursive_ccs.rccs import RecursiveCCS
+from elk.extensions.recursive_ccs.parser import get_rccs_training_parser
+
+
+@torch.autocast("cuda", enabled=False)
```
Comment (norabelrose) [inline elk/extensions/recursive_ccs/train.py:18]: Should just be removed; there's no outer context that's turning `autocast` on so we don't need to explicitly turn it off here

File: elk/training/ccs.py [hunk 3]
```diff
@@ -29,6 +30,7 @@ def __init__(
         loss: Literal["js", "squared"] = "squared",
         num_layers: int = 1,
         pre_ln: bool = False,
+        first_linear_parametrization: Optional[nn.Module] = None,
```
Comment (norabelrose) [inline elk/training/ccs.py:33]: I would prefer to have `RecursiveCCS` just directly call `P.register_parametrization` on `.probe[0]` of the appropriate `CCS` object, instead of cluttering `CCS.__init__` with another parameter

File: elk/extensions/recursive_ccs/parser.py [hunk 4]
```diff
@@ -0,0 +1,12 @@
+from elk.training.parser import get_training_parser
+from argparse import ArgumentParser
+
+
+def get_rccs_training_parser() -> ArgumentParser:
+    parser = get_training_parser()
+    parser.add_argument(
+        "--num-recursive-iterations",
```
Comment (norabelrose) [inline elk/extensions/recursive_ccs/parser.py:8]: Could be shortened to `--num-iterations` or `--num-iter` or `--num-directions`

File: elk/extensions/recursive_ccs/train.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
+import csv
```
Comment (norabelrose) [inline elk/extensions/recursive_ccs/train.py:1]: It's not clear why this entire file is necessary. It seems to be mostly copied over from the primary `train.py`. Could we add a flag or a subparser to the main command instead?

**Review by norabelrose — CHANGES_REQUESTED** [review]

File: elk/extensions/recursive_ccs/parser.py (hunk 4 above)
Comment (FabienRoger) [reply elk/extensions/recursive_ccs/parser.py:8]: Done!

File: elk/training/ccs.py (hunk 3 above)
Comment (FabienRoger) [reply elk/training/ccs.py:33]: Done!

File: elk/extensions/recursive_ccs/train.py (hunk 2 above)
Comment (FabienRoger) [reply elk/extensions/recursive_ccs/train.py:18]: Done

Comment (FabienRoger) [reply elk/extensions/recursive_ccs/train.py]: Also, torch's documentation advises against using autocast for training. Maybe this should be disabled in the regular train.py?

Comment (norabelrose) [reply elk/extensions/recursive_ccs/train.py]: Um, that's just wrong? Autocast is frequently used for training. You can run into problems if you don't use `GradScaler` though

Comment (FabienRoger) [reply elk/extensions/recursive_ccs/train.py]: But we don't use GradScaler, right?

File: elk/training/ccs.py (hunk 1 above)
Comment (norabelrose) [reply elk/training/ccs.py:112]: If you just call `.float()` on the input before calling `fit` this should work

File: elk/extensions/recursive_ccs/train.py (hunk 2 above)
Comment (norabelrose) [reply elk/extensions/recursive_ccs/train.py]: Yep that's why I'm getting rid of autocast in the `semi-supervised` branch. LBFGS currently isn't compatible with `GradScaler` annoyingly. In the future I might try to manually add grad scaling to LBFGS
